### ISSUE TITLE
Ivy: fullTemplateTypeCheck never completes

### ISSUE THREAD

**Issue by mattlewis92** [angular/angular#33532]
# 🐞 bug report

### Affected Package
The issue is caused by package @angular/compiler

### Is this a regression?
Yes, the previous version in which this bug was not present was 8.x

### Description
When testing 9.x on our app it never completed with `fullTemplateTypeCheck` enabled. I managed to track it down to a single component causing the issue, it's very weird, I tried deleting random bits of the template until it worked and there didn't seem to be any consistency to what triggered it to hang. Maybe it's to do with the length of the component?

## 🔬 Minimal Reproduction
```
git clone git@github.com:mattlewis92/angular-ivy-template-check-issue.git
cd angular-ivy-template-check-issue
npm i
npm start
```

and then you'll see it get stuck on `10% building 3/3 modules 0 active`

## 🔥 Exception or Error
N/A

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 9.0.0-rc.0
Node: 12.10.0
OS: darwin x64
Angular: 9.0.0-rc.0
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.0-rc.0
@angular-devkit/build-angular     0.900.0-rc.0
@angular-devkit/build-optimizer   0.900.0-rc.0
@angular-devkit/build-webpack     0.900.0-rc.0
@angular-devkit/core              9.0.0-rc.0
@angular-devkit/schematics        9.0.0-rc.0
@ngtools/webpack                  9.0.0-rc.0
@schematics/angular               9.0.0-rc.0
@schematics/update                0.900.0-rc.0
rxjs                              6.5.3
typescript                        3.6.4
webpack                           4.41.2
</code></pre>

**Anything else relevant?**
Nope

**Comment by JoostK** [comment]
Hi Matt!

You appear to have found a state explosion in TypeScript's type narrowing mechanism!

I was able to reduce the reproduction to the following:

```html
<div *ngIf="planId === freePlan">
  <div *ngIf="planId !== freePlan && planId !== unlimitedPlan"></div>
  <div *ngIf="planId !== freePlan && planId !== unlimitedPlan"></div>
  <div *ngIf="planId !== freePlan && planId !== unlimitedPlan"></div>
  <div *ngIf="planId !== freePlan && planId !== unlimitedPlan"></div>
  <div *ngIf="planId !== freePlan && planId !== unlimitedPlan"></div>
  <div *ngIf="planId !== freePlan && planId !== unlimitedPlan"></div>
  <div *ngIf="planId !== freePlan && planId !== unlimitedPlan"></div>
  <div *ngIf="planId !== freePlan && planId !== unlimitedPlan"></div>
</div>
```

With the following code:

```typescript
import {Component} from '@angular/core';

export enum TeamPlan {
  FREE,
  UNLIMITED,
}

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
})
export class AppComponent {
  freePlan = TeamPlan.FREE;
  unlimitedPlan = TeamPlan.UNLIMITED;
  planId: TeamPlan;
}
```

<details>
 <summary>Generated TCB</summary>

```
import * as i0 from './src/app/app.component';
import * as i1 from '@angular/common';


const _ctor1: (init: Pick<i1.NgIf, "ngIf" | "ngIfThen" | "ngIfElse">) => i1.NgIf = (null!);

/*tcb1*/
function _tcb1(ctx: i0.AppComponent) { if (true) {
    var _t1 = _ctor1({ ngIf: (ctx.planId === ctx.freePlan as any), ngIfThen: (null as any), ngIfElse: (null as any) });
    var _t2: any = (null!);
    if (ctx.planId === ctx.freePlan) {
        var _t3 = document.createElement("div");
        var _t4 = _ctor1({ ngIf: (ctx.planId !== ctx.freePlan && ctx.planId !== ctx.unlimitedPlan as any), ngIfThen: (null as any), ngIfElse: (null as any) });
        var _t5: any = (null!);
        if (ctx.planId !== ctx.freePlan && ctx.planId !== ctx.unlimitedPlan) {
            var _t6 = document.createElement("div");
        }
        var _t7 = _ctor1({ ngIf: (ctx.planId !== ctx.freePlan && ctx.planId !== ctx.unlimitedPlan as any), ngIfThen: (null as any), ngIfElse: (null as any) });
        var _t8: any = (null!);
        if (ctx.planId !== ctx.freePlan && ctx.planId !== ctx.unlimitedPlan) {
            var _t9 = document.createElement("div");
        }
        var _t10 = _ctor1({ ngIf: (ctx.planId !== ctx.freePlan && ctx.planId !== ctx.unlimitedPlan as any), ngIfThen: (null as any), ngIfElse: (null as any) });
        var _t11: any = (null!);
        if (ctx.planId !== ctx.freePlan && ctx.planId !== ctx.unlimitedPlan) {
            var _t12 = document.createElement("div");
        }
        var _t13 = _ctor1({ ngIf: (ctx.planId !== ctx.freePlan && ctx.planId !== ctx.unlimitedPlan as any), ngIfThen: (null as any), ngIfElse: (null as any) });
        var _t14: any = (null!);
        if (ctx.planId !== ctx.freePlan && ctx.planId !== ctx.unlimitedPlan) {
            var _t15 = document.createElement("div");
        }
        var _t16 = _ctor1({ ngIf: (ctx.planId !== ctx.freePlan && ctx.planId !== ctx.unlimitedPlan as any), ngIfThen: (null as any), ngIfElse: (null as any) });
        var _t17: any = (null!);
        if (ctx.planId !== ctx.freePlan && ctx.planId !== ctx.unlimitedPlan) {
            var _t18 = document.createElement("div");
        }
        var _t19 = _ctor1({ ngIf: (ctx.planId !== ctx.freePlan && ctx.planId !== ctx.unlimitedPlan as any), ngIfThen: (null as any), ngIfElse: (null as any) });
        var _t20: any = (null!);
        if (ctx.planId !== ctx.freePlan && ctx.planId !== ctx.unlimitedPlan) {
            var _t21 = document.createElement("div");
        }
        var _t22 = _ctor1({ ngIf: (ctx.planId !== ctx.freePlan && ctx.planId !== ctx.unlimitedPlan as any), ngIfThen: (null as any), ngIfElse: (null as any) });
        var _t23: any = (null!);
        if (ctx.planId !== ctx.freePlan && ctx.planId !== ctx.unlimitedPlan) {
            var _t24 = document.createElement("div");
        }
        var _t25 = _ctor1({ ngIf: (ctx.planId !== ctx.freePlan && ctx.planId !== ctx.unlimitedPlan as any), ngIfThen: (null as any), ngIfElse: (null as any) });
        var _t26: any = (null!);
        if (ctx.planId !== ctx.freePlan && ctx.planId !== ctx.unlimitedPlan) {
            var _t27 = document.createElement("div");
        }
    }
} }
```
</details>

The type checking of this TCB takes **14 seconds** for me, adding an additional `<div *ngIf="planId !== freePlan && planId !== unlimitedPlan"></div>` brings this up to **62 seconds** on my laptop.

I thought that reducing the type of `AppComponent.freePlan` and `AppComponent.unlimitedPlan` as follows could perhaps reduce the time, but it becomes worse at a hefty **941 seconds**, or just over 15 minutes:

```typescript
  freePlan = TeamPlan.FREE as const;
  unlimitedPlan = TeamPlan.UNLIMITED as const;
```

which behaves identical to:

```typescript
  freePlan: TeamPlan.FREE = TeamPlan.FREE;
  unlimitedPlan: TeamPlan.UNLIMITED = TeamPlan.UNLIMITED;
```

Changing it to:

```typescript
  freePlan: unknown = TeamPlan.FREE;
  unlimitedPlan: unknown = TeamPlan.UNLIMITED;
```

finishes in **194 seconds**, or just over 3 minutes. Only changing the type to become `any` resolves the issue:

```typescript
  freePlan: any = TeamPlan.FREE;
  unlimitedPlan: any = TeamPlan.UNLIMITED;
```

This completes in 90 milliseconds on my laptop.

This is ultimately something that should be reported with the TypeScript team, as from Angular's perspective I don't think we can alleviate this issue.

**Comment by mattlewis92** [comment]
Wow, that's crazy, thanks so much for looking into that so quickly, I really appreciate it! ❤️  I can confirm that the `:any` casting works great as a workaround, I'm not familiar enough with how any of this works to feel comfortable filing a bug with TS myself, but I think it's something that should be done as I guess other people will probably end up hitting the same issue once v9 is released.

**Comment by Splaktar** [comment]
Great catch! @JoostK can you please open an issue against TypeScript and link it here?

**Comment by JoostK** [comment]
Opened an issue for the TypeScript team: https://github.com/microsoft/TypeScript/issues/35205

**Comment by nickbullock** [comment]
Hi @IgorMinar , Typescript's issue was closed, is there any news about fullTemplateTypeCheck?

**Comment by JoostK** [comment]
With support for TypeScript 3.8 now included in master and slated for the Angular 9.1 release, I consider this issue to have been resolved.
